### ISSUE TITLE
Support Google Vertex

### ISSUE THREAD

**Issue by cashpw** [karthink/gptel#220]
Google Vertex provides easy access to [many models](https://console.cloud.google.com/vertex-ai/model-garden). I think it'd be okay to reference [`llm-vertex.el`](https://github.com/ahyatt/llm/blob/4add6102f83176beb9655083dd3844ab91672238/llm-vertex.el) as both projects have the same GPL 3.0 license.

**Comment by karthink** [comment]
@cashpw Thank you.  This (and your Google Gemini/Vertex related PRs and issues) are on the docket, I'll get to them soon.

**Comment by algal** [comment]
I may have time to implement this, but first I'm trying to understand what would be involved.

If I understand right, llm-vertex.el does not actually have a working emacs lisp implementation which can authenticate and interact with vertex. Instead, they effectively require you to install the gcloud binary and configure it. That does not seem idle.

Do we know if anyone has a pure emacs lisp client that can authenticate with vertex ai?

**Comment by karthink** [comment]
@algal I haven't had time to look into this.  It looks like Vertex AI [supports a REST interface](https://cloud.google.com/vertex-ai/docs/reference/rest) though.  The `generateContent` endpoint is the one that gptel would support.

**Comment by algal** [comment]
Here's a quick update on this:

I just had a look at llm-vertex.el, and was able to use it to query VertexAI with authentication successfully. This required setting up the gcloud binary and authenticating with it. The lisp code for making a request looks like this, fyi:

```lisp
(customize-set-value 'llm-vertex-gcloud-binary "/Users/alexis/miniforge3/bin/gcloud")
(customize-set-value 'llm-vertex-gcloud-region "us-east5")
(llm-chat (make-llm-vertex
           :project "PROJECT_ID_HERE"
           :chat-model "gemini-pro"
           :default-chat-temperature 0.5
           :default-chat-max-tokens 1000
           :default-chat-non-standard-params (list))
          (llm-make-chat-prompt
           "Hello, are you there?"))
```

I also had a look at the extensive (and imho unnecessarily confusing) list of Vertex AI endpoints. 

As far as I can tell:
- the `generateContent` endpoint is the one which corresponds to the "metric" described in the Gcloud API as "aiplatform.googleapis.com/generate_content_requests_per_minute_per_project_per_base_model". 
- that endpoint or metric offers various Gemini models and some Llama models, but it does *not* offer any anthropic models.
- this is in contrast to "Online prediction requests per base model per minute per region per base_model" item in the GCloud UI, which *does* offer the latest Anthropic model. I believe that item corresponds to the `streamRawPredict` endpoint which [Anthropic shows in their example code](https://docs.anthropic.com/en/api/claude-on-vertex-ai).
- therefore, it's not possible to use `generateContent` (llm-vertex.el with its defualts) to access Anthropic models via VertexAI.

That's my main use case, so it seems like the relatively easy paths to adding vertexai support don't meet my needs, so I'm not planning further work on this at the moment. But I'll keep an eye on this issue in case the situation changes.

**Comment by d-miketa** [comment]
I would also be interested in support for Vertex.

**Comment by ashlineldridge** [comment]
I have created a separate package [gptel-vertex](https://github.com/ashlineldridge/gptel-vertex) that provides basic Vertex support for gptel. The package supports Gemini and Claude models and it should be straightforward to add other models. Note that I authenticate with GCP using ADC credentials (i.e. `gcloud auth login --update-adc`).

Models supported/tested:
- `gemini-2.5-flash`
- `gemini-2.5-pro`
- `claude-sonnet-4-5@20250929`
- `claude-opus-4-1@20250805`

If there is interest, and after I develop more confidence in my implementation, I'm open to raising a PR to upstream this into `gptel`.

**Comment by karthink** [comment]
Non-substantive update: Vertex support is still on the docket, i've just been stretched thin.  I'll take a look at @ashlineldridge's implementation soon, and see if we can use that as a base.  I'll probably have to work on adding general/reusable Oauth flow support to gptel first, though, since many backends implement their own right now.

**Comment by berenddeboer** [comment]
I've tried this, however I get: `VertexAI error: (HTTP/2 404) Malformed JSON in response.`

**Comment by dodjiseketeli** [comment]
Hello, just to let you know that I have successfully been testing the code from @ashlineldridge at https://github.com/ashlineldridge/gptel-vertex and it works great.  Many thanks to @ashlineldridge for the great work.
